go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go

```diff
@@ -286,44 +306,99 @@ func newTxThrottlerState(topoServer *topo.Server, config *txThrottlerConfig, tar
 		t.Close()
 		return nil, err
 	}
-	result := &txThrottlerState{
+	state := &txThrottlerState{
 		config:    config,
 		throttler: t,
 	}
-	createTxThrottlerHealthCheck(topoServer, config, result, target.Cell)
 
-	result.topologyWatchers = make(
-		[]TopologyWatcherInterface, 0, len(config.healthCheckCells))
-	for _, cell := range config.healthCheckCells {
-		result.topologyWatchers = append(
-			result.topologyWatchers,
+	// get cells from topo if none defined in tabletenv config
+	if len(config.healthCheckCells) == 0 {
+		var err error
+		if config.healthCheckCells, err = fetchKnownCells(topoServer); err != nil {
+			log.Errorf("txThrottler: failed to open throttler: %+v", err)
+			return nil, err
+		}
+		state.cellsFromTopo = true
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	state.stopHealthCheck = cancel
+	state.initHealthCheckStream(topoServer, target)
+	hcProcessor := state.healthChecksProcessorFactory(topoServer, target)
+	go hcProcessor(ctx)
+
+	return state, nil
+}
+
+func (ts *txThrottlerState) initHealthCheckStream(topoServer *topo.Server, target *querypb.Target) {
+	ts.healthCheck = healthCheckFactory(topoServer, target.Cell, ts.config.healthCheckCells)
+	ts.healthCheckChan = ts.healthCheck.Subscribe()
+	ts.topologyWatchers = make(
+		[]TopologyWatcherInterface, 0, len(ts.config.healthCheckCells))
+	for _, cell := range ts.config.healthCheckCells {
+		ts.topologyWatchers = append(
+			ts.topologyWatchers,
 			topologyWatcherFactory(
 				topoServer,
-				result.healthCheck,
+				ts.healthCheck,
 				cell,
 				target.Keyspace,
 				target.Shard,
 				discovery.DefaultTopologyWatcherRefreshInterval,
-				discovery.DefaultTopoReadConcurrency))
+				discovery.DefaultTopoReadConcurrency),
+		)
 	}
-	return result, nil
 }
 
-func createTxThrottlerHealthCheck(topoServer *topo.Server, config *txThrottlerConfig, result *txThrottlerState, cell string) {
-	ctx, cancel := context.WithCancel(context.Background())
-	result.stopHealthCheck = cancel
-	result.healthCheck = healthCheckFactory(topoServer, cell, config.healthCheckCells)
-	ch := result.healthCheck.Subscribe()
-	go func(ctx context.Context) {
+func (ts *txThrottlerState) closeHealthCheckStream() {
+	if ts.healthCheck == nil {
+		return
+	}
+	for _, watcher := range ts.topologyWatchers {
+		watcher.Stop()
+	}
+	ts.topologyWatchers = nil
+	ts.stopHealthCheck()
+	ts.healthCheck.Close()
+}
+
+func (ts *txThrottlerState) healthChecksProcessorFactory(topoServer *topo.Server, target *querypb.Target) func(ctx context.Context) {
+	if ts.cellsFromTopo {
+		return func(ctx context.Context) {
+			cellsUpdateTicker := time.NewTicker(topoCellsRefreshInterval)
+			defer cellsUpdateTicker.Stop()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-cellsUpdateTicker.C:
+					cells, err := fetchKnownCells(topoServer)
```

Comment:
please pass `ctx` to `fetchKnownCells()` so that if `ctx` is cancelled, that breaks out of `fetchKnownCells()` sooner rather than later.